src/__tests__/types.typescript.tsx

```diff
@@ -0,0 +1,325 @@
+
+const baseReturn = typeBaseHook();
+
+// @ts-ignore Should not accept arguments
+typeBaseHook({});
+
+// @ts-ignore Array index 0 should be state
+baseReturn[0].foo;
+
+// @ts-ignore Array index 1 should be actions
+baseReturn[1].increment();
+
+// @ts-ignore Array index 1 should be actions
+baseReturn[1].increment("1");
+
+// @ts-ignore Array index 1 return type should be correct
+baseReturn[1].decrement().then(v => v);
+
+// Correct
+baseReturn[0].count + 0;
+baseReturn[1].increment(1);
+baseReturn[1].decrement();
+baseReturn[1].fetch().then(v => v);
+
+/**
+ * createHook with selector types tests
+ */
+const typeSelectorHook = createHook<State, Actions, { baz: number }>(
+  TypeStore,
+  { selector: state => ({ baz: 1 }) }
+);
+
+const selectorReturn = typeSelectorHook();
+
+// @ts-ignore Array index 0 shape should be selector output
+selectorReturn[0].count;
+
+// @ts-ignore Should not accept props
+typeSelectorHook({ min: 3 });
+
+// Correct
+selectorReturn[0].baz;
+selectorReturn[1].increment(1);
+
+const typeNullHook = createHook<State, Actions, void>(TypeStore, {
+  selector: null
+});
+
+const nullReturn = typeNullHook();
+
+// @ts-ignore Array 0 shape should be undefined
+nullReturn[0].count;
+
+// Correct
+nullReturn[1].increment(1);
+
+const typeArgHook = createHook<State, Actions, SelectorState, SelectorProps>(
+  TypeStore,
+  { selector: (state, props) => ({ baz: 1, min: props.min }) }
+);
+
+// @ts-ignore Should require argument
+typeArgHook();
+
+// @ts-ignore Should require correct prop types
+typeArgHook({ min: "2" });
+
+const argReturn = typeArgHook({ min: 2 });
+// @ts-ignore Should have correct selector types
+argReturn[0].min.split("");
+
+// Correct
+argReturn[0].min + argReturn[0].baz;
+
+/**
+ * Container types tests
+ */
+
+const TypeContainer = createContainer<State, Actions>(TypeStore);
+
+Test = (
+  // @ts-ignore Container is not a render-prop
+  <TypeContainer>{({ count }) => count}</TypeContainer>
+);
+
+Test = (
+  // @ts-ignore Does not accept extra props
+  <TypeContainer foo="1">bla</TypeContainer>
+);
+
+// Correct
+Test = <TypeContainer>bla</TypeContainer>;
+Test = <TypeContainer scope="a">bla</TypeContainer>;
+Test = <TypeContainer isGlobal>bla</TypeContainer>;
+
+const TypePropsContainer = createContainer<State, Actions, { url: string }>(
+  TypeStore
+);
+
+// @ts-ignore Requires typed props
+Test = <TypePropsContainer isGlobal>bla</TypePropsContainer>;
+
+Test = (
+  // @ts-ignore Only allows typed extra props
+  <TypePropsContainer foo="1">bla</TypePropsContainer>
+);
+
+// Correct
+Test = (
+  <TypePropsContainer scope="a" url="">
+    bla
+  </TypePropsContainer>
+);
+
+
+const inferedStore = createStore({
+  initialState: {
+    counter : 0
+  },
+  actions: {
+    increment: (by: number = 1) => ({getState, setState}) =>{
+      setState({
+        counter : getState().counter + by
+      })
+    },
+    composite : () => ({actions}) => {
+      actions.increment(2);
```

Comment:
Actions type is resolved as `unknown` in that case and compilation fails.

Reply:
Decided to leave `actions` out and deprecate that pattern due to typing complications